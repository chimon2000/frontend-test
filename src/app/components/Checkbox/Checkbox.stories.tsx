import * as React from "react";
import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import { action } from "@storybook/addon-actions";
import { boolean, text } from "@storybook/addon-knobs";

import { Checkbox } from "./Checkbox";

const stories = storiesOf("Components", module);

stories.add(
  "Checkbox",
  withInfo({ inline: true })(() => (
    <Checkbox
      onChange={action("Change event")}
      checked={boolean("Checked", true)}
      disabled={boolean("Disabled", false)}
      name="checkbox_test"
      id="checkbox_test"
      label="Hello Checkbox"
    />
  ))
);
