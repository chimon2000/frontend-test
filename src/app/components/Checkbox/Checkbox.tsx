import * as React from "react";

import { joinClassList } from "../../helpers";

export interface ChangeEventObject {
  name: string;
  value: boolean;
}
export interface CheckboxProps {
  /**
   * Triggers when Checkbox is toggled checked/unchecked.
   */
  onChange: (result: ChangeEventObject) => void;
  /**
   * The Name of the checkbox. Make sure it's unique to the page!
   */
  name: string;
  checked: boolean;
  disabled?: boolean;
  /**
   * Can be text or a React Element.
   */
  label: React.ReactNode;
  children?: React.ReactNode;
  /** Required for inputs. */
  id: string;
  className?: string;
}

export const Checkbox: React.FunctionComponent<CheckboxProps> = ({
  onChange,
  name,
  checked,
  disabled = false,
  label,
  children,
  id,
  className
}) => {
  const [isChecked, toggleChecked] = React.useState(checked);
  const classes = [
    `checkbox`,
    `${className || ""}`,
    `${checked ? "checkbox--checked" : ""}`
  ];

  React.useEffect(() => toggleChecked(checked), [checked]);

  return (
    <div className={joinClassList(classes)}>
      <input
        className="checkbox__input"
        type="checkbox"
        name={name}
        id={id}
        checked={isChecked}
        disabled={disabled}
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
          toggleChecked(!isChecked);
          onChange({ name, value: !isChecked });
        }}
      />
      <label className="checkbox__label" htmlFor={id}>
        {label}
      </label>
    </div>
  );
};
