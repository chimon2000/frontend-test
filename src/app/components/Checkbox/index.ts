export {
  Checkbox as default,
  CheckboxProps,
  ChangeEventObject
} from "./Checkbox";
