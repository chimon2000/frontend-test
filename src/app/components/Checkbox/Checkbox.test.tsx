import * as React from "react";
import { shallow } from "enzyme";

import { Checkbox } from "./Checkbox";

describe("<Checkbox />", () => {
  it("doesn't blow up", () => {
    const wrapper = shallow(
      <Checkbox
        onChange={() => {}}
        checked={true}
        disabled={false}
        label="label"
        name="checkbox_test"
        id="checkbox_test"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("Triggers onChange when change happens", () => {
    let triggered = false;
    const updateMethod = () => {
      triggered = true;
    };
    const wrapper = shallow(
      <Checkbox
        onChange={updateMethod}
        checked={false}
        disabled={false}
        label="label"
        name="checkbox_test"
        id="checkbox_test"
      />
    );

    const input = wrapper.find(".checkbox__input");
    input.simulate("change", { target: { checked: true } });
    expect(triggered).toEqual(true);
  });

  it("Is checked if value is passed", () => {
    const wrapper = shallow(
      <Checkbox
        onChange={() => {}}
        checked={true}
        disabled={false}
        label="label"
        name="checkbox_test"
        id="checkbox_test"
      />
    );
    const input = wrapper.find(".checkbox__input").get(0);
    expect(input.props.checked).toBe(true);
  });
});
