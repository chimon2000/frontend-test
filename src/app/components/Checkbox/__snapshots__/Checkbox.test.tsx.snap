// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> doesn't blow up 1`] = `
<div
  className="checkbox checkbox--checked"
>
  <input
    checked={true}
    className="checkbox__input"
    disabled={false}
    id="checkbox_test"
    name="checkbox_test"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="checkbox__label"
    htmlFor="checkbox_test"
  >
    label
  </label>
</div>
`;
